planning/MartenV4.md

```diff
@@ -106,9 +144,262 @@ This is also covered by the [Linq Overhaul](https://github.com/JasperFx/marten/i
+        // of events. 
+        public void Partial(Event1, @event, IDocumentSession session);
+
+        // Should the aggregate document be deleted based on some event data?
+        public bool ShouldBeDeleted(Event1 @event);
+        public bool ShouldBeDeleted(EventWrapper @event);
+        public bool ShouldBeDeleted(EventWrapper @event, MyEventStream stream);
+    }
+```
+
+#### Defining a Projection to a SQL Statement(s)
+
+TODO
+
+#### Inline Projections
+
+There's been some complaints and obvious confusion about the `IEvent` interface and how the metadata properties like
+`version` aren't available during inline projections. That's because today the inline projections are all calculated 
+just prior to submitting the database commands before the version numbers of each event within the stream would be calculated. 
+
+In v4, the inline projection support and the `UpdateBatch` mechanics should be smart enough to "know" when the transformation
+or aggregation requires metadata, so that:
+
+#### If no event or stream metadata is necessary...
+
+It works basically as it does today. Upon `IDocumentSession.SaveChanges() / SaveChangesAsync()`:
+
+1. The inline projections work across the incoming event objects, create new projected documents
+or update existing documents as necessary just prior to sending the current `UnitOfWork` to the database
+1. The `DocumentSession` issues the batched up commands with both the event capture and the projection documents
+1. The underlying database transaction is committed
+
+If there's no need for event metadata, do thing in fewer network round trips
+
+#### If event or stream metadata is necessary...
+
+This gets a little more complicated, because the projections can't be calculated until *after* the events are appended, so:
+
+1. The events get appended, and the sproc returns an array of the new event versions
+1. Build up an array of `IEvent` wrappers that have the stream and event metadata
+1. Pass the `IEvent` wrappers into the transformation code, and get an additional `UpdateBatch` of projected document inserts
+   and updates
+1. Submit the subsequent `UpdateBatch` to the database
+1. The underlying database transaction is committed
+
+The difference here is that this will take multiple batched commands to the database within the same `NpgsqlTransaction`
+
+
+
+
+
+### Snapshots & "Live" Projections
+
+
+## Async Daemon
+
+For v4, the Async Daemon needs to come in a couple different possible deployment configurations:
+
+1. "Classic" database polling mode -- no queueing, just something that polls for new events and runs the projections. I think this will be a near rewrite in v4, but there's some 
+   existing art in [Jasper's DurabilityAgent](http://jasperfx.github.io/documentation/durability/) that might help this along
+1. Messaging-based Mode -- use Rabbit MQ, Azure Service Bus, Kafka, etc. to gather up events and connect with the running projections
```

Comment:
This is...I mean, I would love to have this. Dont get me wrong. But there are several things to take into account here. Ill come back to this later when Im done iwth my writeup.